### PR TITLE
Add `LazyJSONDocument`, which wraps a JSON string and only deserializes it if needed.

### PR DESCRIPTION
**Pull request by nicktobey**
This is the GMS side of https://github.com/dolthub/dolt/issues/7749

This is a new `JSONWrapper` implementation. It isn't used by the GMS in-memory storage, but it will be used in Dolt to speed up `SELECT` queries that don't care about the structure of the JSON.

A big difference between this and `JSONDocument` is that even after it de-serializes the JSON into a go value, it continues to keep the string in memory. This is good in cases where we would want to re-serialize the JSON later without changing it. (So statements like `SELECT json FROM table WHERE json->>"$.key" =  "foo";` will still be faster.) But with the downside of using more memory than `JSONDocument`)

### CODE REVIEW

**Comment by nicktobey** [comment]
One detail to note is that we currently desalinize JSON prior to writing it on the wire, so that we can re-serialize it in a way that matches MySQL's output (different whitespacing.) This is expensive and I don't think is worth it. 

I discussed with Max whether or not we consider it a requirement to match MySQL's wire responses exactly for JSON, and agreed that we could probably relax that requirement. With this change, casting a document to a text type will still result in the same output as MySQL, but simply displaying it in a query result won't.

File: sql/types/json_value.go [hunk 1]
```diff
@@ -82,19 +119,50 @@ func (doc JSONDocument) String() string {
...
+		return nil, err
+	}
+	return ContainsJSON(doc.Val, candidateVal)
 }
 
 func (doc JSONDocument) Extract(path string) (sql.JSONWrapper, error) {
 	return LookupJSONValue(doc, path)
 }
 
+type LazyJSONDocument struct {
```
Comment (fulghum) [inline sql/types/json_value.go:136]: Would be helpful to document this type and the intent for why it exists.

**Review by fulghum — APPROVED** [review]
Looks good. It seems reasonable to tradeoff fully matching the whitespace formatting in MySQL's JSON output, with getting faster performance.

File: sql/types/json_test.go [hunk 2]
```diff
@@ -212,6 +212,29 @@ func TestValuer(t *testing.T) {
 	require.Equal(t, `{"a": "one"}`, res)
 }
 
+func TestLazyJsonDocument(t *testing.T) {
+	testCases := []struct {
+		s    string
+		json interface{}
+	}{
+		{"\"1\"", "1"},
+		{"{\"a\": [1.0, null]}", map[string]any{"a": []any{1.0, nil}}},
```
Comment (fulghum) [inline sql/types/json_test.go:221]: (super minor) backtick quoted strings in Go are handy for avoiding having to escape double quotes. 

```suggestion
		{`"1"`, "1"},
		{`{"a": [1.0, null]}`, map[string]any{"a": []any{1.0, nil}}},
```

Comment (nicktobey) [reply sql/types/json_test.go]: Done.

**Pull request merged**